interceptor.go

```diff
@@ -142,78 +63,66 @@ func (i *interceptor) WrapUnary(next connect.UnaryFunc) connect.UnaryFunc {
 				return next(ctx, request)
 			}
 		}
-		tracer := i.config.Tracer()
-		carrier := propagation.HeaderCarrier(request.Header())
-		attrs := attributesFromRequest(req)
+		isClient := request.Spec().IsClient
 		name := strings.TrimLeft(request.Spec().Procedure, "/")
-		protocol := protocolToSemConv(request.Peer())
-		var span trace.Span
-		if request.Spec().IsClient {
-			ctx, span = tracer.Start(
-				ctx,
-				name,
-				trace.WithSpanKind(trace.SpanKindClient),
-				trace.WithAttributes(attrs...),
-			)
-		} else {
-			ctx = i.config.propagator.Extract(ctx, carrier)
-			spanCtx := trace.SpanContextFromContext(ctx)
-			ctx, span = tracer.Start(
-				trace.ContextWithRemoteSpanContext(ctx, spanCtx),
-				name,
-				trace.WithSpanKind(trace.SpanKindServer),
-				trace.WithAttributes(attrs...),
-			)
+		protocol := protocolToSemConv(request.Peer().Protocol)
+		attributes := requestAttributes(req)
+		instrumentation, err := i.getAndInitInstrument(isClient)
+		if err != nil {
+			return nil, connect.NewError(connect.CodeInternal, err)
+		}
+		spanKind := trace.SpanKindClient
+		requestSpan, responseSpan := semconv.MessageTypeSent, semconv.MessageTypeReceived
+		if !isClient {
+			spanKind = trace.SpanKindServer
+			requestSpan, responseSpan = responseSpan, requestSpan
 		}
+		carrier := propagation.HeaderCarrier(request.Header())
+		spanCtx := trace.SpanContextFromContext(ctx)
+		ctx = trace.ContextWithRemoteSpanContext(ctx, spanCtx)
+		ctx = i.config.propagator.Extract(ctx, carrier)
+		ctx, span := i.config.tracer.Start(
+			ctx,
+			name,
+			trace.WithSpanKind(spanKind),
+		)
```

Comment:
All of the context handling is now the same for clients + servers

Reply:
Resolved in ebe6bd9ccb0038f11b8b48d1dd98ea5abe9e930f and in 43c4f2f4e8f0e501af4fe9c3b2d93dbc72dd227a

Tested in 43a064d5a2b4212aa8f2dbf7095aeb35fe4f3d0f